gluten-core/src/main/java/io/glutenproject/memory/memtarget/spark/TreeMemoryConsumer.java

```diff
@@ -0,0 +1,307 @@
+ * free out enough memory, throw OOM to caller.
+ *
+ * <p>Spark's memory manager could either trigger spilling on the children spillers since this was
+ * registered as a Spark memory consumer.
+ *
+ * <p>Typically used by utility class {@link
+ * io.glutenproject.memory.memtarget.spark.IsolatedMemoryConsumers}.
+ */
+public class TreeMemoryConsumer extends MemoryConsumer implements TreeMemoryConsumerNode {
+  private final SimpleMemoryUsageRecorder recorder = new SimpleMemoryUsageRecorder();
+  private final Map<String, TreeMemoryConsumerNode> children = new HashMap<>();
+
+  TreeMemoryConsumer(TaskMemoryManager taskMemoryManager) {
+    super(taskMemoryManager, taskMemoryManager.pageSizeBytes(), MemoryMode.OFF_HEAP);
+  }
+
+  @Override
+  public long borrow(long size) {
+    if (size == 0) {
+      // or Spark complains about the zero size by throwing an error
+      return 0;
+    }
+    long acquired = acquireMemory(size);
+    recorder.inc(acquired);
+    return acquired;
+  }
+
+  @Override
+  public long repay(long size) {
+    if (size == 0) {
+      return 0;
+    }
+    freeMemory(size);
+    Preconditions.checkArgument(getUsed() >= 0);
+    recorder.inc(-size);
+    return size;
+  }
+
+  @Override
+  public String name() {
+    return "Gluten.Tree@" + Integer.toHexString(System.identityHashCode(this));
+  }
+
+  @Override
+  public long usedBytes() {
+    return getUsed();
+  }
+
+  @Override
+  public MemoryUsageStats stats() {
+    Set<Map.Entry<String, TreeMemoryConsumerNode>> entries = children.entrySet();
+    Map<String, MemoryUsageStats> childrenStats =
+        entries.stream()
+            .collect(Collectors.toMap(e -> e.getValue().name(), e -> e.getValue().stats()));
+
+    Preconditions.checkState(childrenStats.size() == children.size());
+    MemoryUsageStats stats = recorder.toStats(childrenStats);
+    Preconditions.checkState(
+        stats.getCurrent() == getUsed(),
+        "Used bytes mismatch between gluten memory consumer and Spark task memory manager");
+    return stats;
+  }
+
+  @Override
+  public TaskMemoryManager getTaskMemoryManager() {
+    return taskMemoryManager;
+  }
+
+  @Override
+  public long spill(long size, MemoryConsumer trigger) throws IOException {
+    // subject to the regular Spark spill calls
+    return spillTree(this, size);
+  }
+
+  static long spillTree(TreeMemoryConsumerNode node, final long bytes) {
+    // sort children by used bytes, descending
+    Queue<TreeMemoryConsumerNode> q =
+        new PriorityQueue<>(
+            (o1, o2) -> {
+              long diff = o1.usedBytes() - o2.usedBytes();
+              return -(diff > 0 ? 1 : diff < 0 ? -1 : 0); // descending
+            });
+    q.addAll(node.children().values());
+
+    long remainingBytes = bytes;
+    while (q.peek() != null && remainingBytes > 0) {
+      TreeMemoryConsumerNode head = q.remove();
+      long spilled = spillTree(head, remainingBytes);
+      remainingBytes -= spilled;
+    }
```

Comment:
This logic wants invoke spill from smallest consumer to largest consumer right?

Reply:
It's from largest to smallest.

In future we may want to follow vanilla Spark's rule, which uses the smallest one among those are larger than the target size.

Reply:
Still not fully understanding...

We sort children in descending order and invoke spillTree on peek element's children recursively, which means we pick largest consumer and then pass smaller consumer into spillTree, when node has no children, we spill it and return.
I think spillTree loop from largest to smallest, but spill from smallest to largest, please correct me if I'm wrong, thanks!

Another misunderstanding logic, does the sort of root not sort its all children? seems  we sort node's children in every spillTree invoked.

Reply:
Hi, thanks for helping checking this code. I haven't verified it carefully but let's take a simple example:
```
a 200 (self 80)
|-b 70
|-c 50 (self 10)
   |- d 30
   |- e 10
```

With the code we implemented post-order traversal on this tree (children-first, self-last), which means the visiting order is supposed to be

```
b (70) -> d (30) -> e (10) -> c (10) -> a (80)
```

Which seems to be aligned with my initial assumption: **largest to smallest, but self last**.
Is that the same with your thoughts?

> Another misunderstanding logic, does the sort of root not sort its all children? seems we sort node's children in every spillTree invoked.

Yes the data structure is not efficient since we only sorted the children of the current node (See code `q.addAll(node.children().values());`). So probably we can move to TreeMap/TreeSet later on.

Reply:
> Which seems to be aligned with my initial assumption: largest to smallest, but self last.
Is that the same with your thoughts?

Thanks, it's same.